assets/js/base/hooks/use-position-relative-to-viewport.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * External dependencies
+ */
+import { useRef, useLayoutEffect, useState } from '@wordpress/element';
+
+/** @type {React.CSSProperties} */
+const style = {
+	bottom: 0,
+	left: 0,
+	opacity: 0,
+	pointerEvents: 'none',
+	position: 'absolute',
+	right: 0,
+	top: 0,
+	zIndex: -1,
+};
+
+/**
+ * Returns an element and a string (`above`, `visible` or `below`) based on the
+ * element position relative to the viewport.
+ * _Note: `usePositionRelativeToViewport` will return an empty position (``)
+ * until after first render_
+ *
+ * @return {Array} An array of {Element} `referenceElement` and {string} `positionRelativeToViewport`.
+ *
+ * @example
+ *
+ * ```js
+ * const App = () => {
+ * 	const [ referenceElement, positionRelativeToViewport ] = useContainerQueries();
+ *
+ * 	return (
+ * 		<>
+ * 			{ referenceElement }
+ * 			{ positionRelativeToViewport === 'below' && <p>Reference element is below the viewport.</p> }
+ * 			{ positionRelativeToViewport === 'visible' && <p>Reference element is visible in the viewport.</p> }
+ * 			{ positionRelativeToViewport === 'above' && <p>Reference element is above the viewport.</p> }
+ * 		</>
+ * 	);
+ * };
+ * ```
+ */
+export const usePositionRelativeToViewport = () => {
+	const [
+		positionRelativeToViewport,
+		setPositionRelativeToViewport,
+	] = useState( '' );
+	const referenceElementRef = useRef( null );
+	const intersectionObserver = useRef(
+		new IntersectionObserver(
+			( entries ) => {
+				if ( entries[ 0 ].isIntersecting ) {
+					setPositionRelativeToViewport( 'visible' );
+				} else {
+					setPositionRelativeToViewport(
+						entries[ 0 ].boundingClientRect.top > 0
+							? 'below'
+							: 'above'
+					);
+				}
+			},
+			{ threshold: 1.0 }
```

Comment:
If I'm not wrong, `root` is expected to be the scrollable ancestor of the observed element. In this case, there is no scrollable ancestor other than the browser window, so it should be `null`.

Just to confirm, I tested setting `root: document.body.querySelector( '.entry-content' )` and it didn't seem to work because `isIntersecting` was always `true`.

Reply:
👍 Thanks for the explanation. That works for me.